File: cuty_app/lib/services/school_service.dart
import 'package:cuty_app/config/app_config.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import '../models/school.dart';
import '../models/college.dart';
import '../models/department.dart';

class SchoolService {
  final String baseUrl = AppConfig.baseApiUrl;

  Future<Map<String, dynamic>> getSchools(
    int countryId, {
    int page = 1,
    int perPage = 10,
    String search = '',
  }) async {
    try {
      final response = await http.get(
        Uri.parse(
            '$baseUrl/countries/$countryId/schools?page=$page&per_page=$perPage&search=$search'),
      );

      if (response.statusCode == 200) {
        print(response.body);
        final Map<String, dynamic> data = json.decode(response.body);
        final List<School> schools = (data['schools'] as List)
            .map((json) => School.fromJson(json))
            .toList();

        return {
          'schools': schools,
          'total': data['total'],
          'pages': data['pages'],
          'currentPage': data['current_page'],
          'perPage': data['per_page'],
          'search': data['search'],
        };
      } else if (response.statusCode == 404) {
        throw Exception('존재하지 않는 국가입니다');
      } else {
        throw Exception('학교 목록을 불러오는데 실패했습니다');
      }
    } catch (e) {
      throw Exception('서버 연결에 실패했습니다: ${e.toString()}');
    }
  }

  Future<Map<String, dynamic>> getColleges(
    int countryId,
    int schoolId, {
    int page = 1,
    int perPage = 10,
    String search = '',
  }) async {
    try {
      final response = await http.get(
        Uri.parse(
            '$baseUrl/countries/$countryId/schools/$schoolId/colleges?page=$page&per_page=$perPage&search=$search'),
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> data = json.decode(response.body);
        final List<College> colleges = (data['colleges'] as List)
            .map((json) => College.fromJson(json))
            .toList();

        return {
          'colleges': colleges,
          'total': data['total'],
          'pages': data['pages'],
          'currentPage': data['current_page'],
          'perPage': data['per_page'],
          'search': data['search'],
        };
      } else if (response.statusCode == 404) {
        throw Exception('존재하지 않는 국가 또는 학교입니다');
      } else if (response.statusCode == 400) {
        throw Exception('해당 국가의 학교가 아닙니다');
      } else {
        throw Exception('단과대학 목록을 불러오는데 실패했습니다');
      }
    } catch (e) {
      print(e);
      throw Exception('서버 연결에 실패했습니다: ${e.toString()}');
    }
  }

  Future<Map<String, dynamic>> getDepartments(
    int countryId,
    int schoolId,
    int collegeId, {
    int page = 1,
    int perPage = 10,
    String search = '',
  }) async {
    try {
      final response = await http.get(
        Uri.parse(
            '$baseUrl/countries/$countryId/schools/$schoolId/colleges/$collegeId/departments?page=$page&per_page=$perPage&search=$search'),
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> data = json.decode(response.body);
        final List<Department> departments = (data['departments'] as List)
            .map((json) => Department.fromJson(json))
            .toList();

        return {
          'departments': departments,
          'total': data['total'],
          'pages': data['pages'],
          'currentPage': data['current_page'],
          'perPage': data['per_page'],
          'search': data['search'],
        };
      } else if (response.statusCode == 404) {
        throw Exception('존재하지 않는 국가, 학교 또는 단과대학입니다');
      } else if (response.statusCode == 400) {
        throw Exception('해당 국가의 학교가 아니거나 해당 학교의 단과대학이 아닙니다');
      } else {
        throw Exception('학과 목록을 불러오는데 실패했습니다');
      }
    } catch (e) {
      throw Exception('서버 연결에 실패했습니다: ${e.toString()}');
    }
  }
}

File: cuty_server/src/services/school_service.py
from src.models import db, Country, School, College, Department
from sqlalchemy import or_
from src.utils.formatters import (
    get_country_data, get_school_data, 
    get_college_data, get_department_data
)

class SchoolService:
    @staticmethod
    def get_schools(page, per_page, search=''):

        schools_query = School.query

        if search:
            schools_query = schools_query.filter(School.name.ilike(f'%{search}%'))
        
        schools_query = schools_query.order_by(School.name.asc())
        pagination = schools_query.paginate(page=page, per_page=per_page, error_out=False)
        
        schools = [get_school_data(school) for school in pagination.items]
        
        return {
            'schools': schools,
            'total': pagination.total,
            'pages': pagination.pages,
            'current_page': page,
            'per_page': per_page,
            'search': search
        }

    @staticmethod
    def get_colleges(school_id, page, per_page, search=''):
        # 학교 존재 여부 확인
        school = School.query.get(school_id)
        if not school:
            raise ValueError('존재하지 않는 학교입니다')

        # 단과대학 쿼리 생성
        colleges_query = College.query.filter_by(school_id=school_id)
        
        if search:
            colleges_query = colleges_query.filter(College.name.ilike(f'%{search}%'))
        
        colleges_query = colleges_query.order_by(College.name.asc())
        pagination = colleges_query.paginate(page=page, per_page=per_page, error_out=False)
        
        colleges = [get_college_data(college) for college in pagination.items]
        
        return {
            'colleges': colleges,
            'total': pagination.total,
            'pages': pagination.pages,
            'current_page': page,
            'per_page': per_page,
            'search': search
        }

    @staticmethod
    def get_departments(school_id, college_id, page, per_page, search=''):
        # 학교 존재 여부 확인
        school = School.query.get(school_id)
        if not school:
            raise ValueError('존재하지 않는 학교입니다')
            
        # 단과대학 존재 여부와 학교 관계 확인
        college = College.query.get(college_id)
        if not college:
            raise ValueError('존재하지 않는 단과대학입니다')
        if college.school_id != school_id:
            raise ValueError('해당 학교의 단과대학이 아닙니다')
        
        # 학과 쿼리 생성
        departments_query = Department.query.filter_by(college_id=college_id)
        
        if search:
            departments_query = departments_query.filter(Department.name.ilike(f'%{search}%'))
        
        departments_query = departments_query.order_by(Department.name.asc())
        pagination = departments_query.paginate(page=page, per_page=per_page, error_out=False)
        
        departments = [get_department_data(department) for department in pagination.items]
        
        return {
            'departments': departments,
            'total': pagination.total,
            'pages': pagination.pages,
            'current_page': page,
            'per_page': per_page,
            'search': search
        }
